src/databricks/labs/ucx/progress/history.py

```diff
@@ -0,0 +1,216 @@
+from __future__ import annotations
+import dataclasses
+import datetime as dt
+import json
+import logging
+from collections.abc import Iterable, Sequence
+from typing import ClassVar, Protocol, TypeVar, Generic, Any, get_type_hints
+
+from databricks.labs.lsql.backends import SqlBackend
+
+from databricks.labs.ucx.framework.owners import Ownership
+from databricks.labs.ucx.framework.utils import escape_sql_identifier
+from databricks.labs.ucx.progress.install import Historical
+
+logger = logging.getLogger(__name__)
+
+
+class DataclassWithIdAttributes(Protocol):
+    __dataclass_fields__: ClassVar[dict[str, Any]]
+
+    __id_attributes__: ClassVar[tuple[str, ...]]
+    """The names of attributes (can be dataclass fields or ordinary properties) that make up the object identifier.
+
+    All attributes must be (non-optional) strings.
+    """
+
+
+Record = TypeVar("Record", bound=DataclassWithIdAttributes)
+
+
+class HistoricalEncoder(Generic[Record]):
+    _job_run_id: int
+    """The identifier of the current job run, with which these records are associated."""
+
+    _workspace_id: int
+    """The identifier of the current workspace, for identifying where these records came from."""
+
+    _ownership: Ownership[Record]
+    """Used to determine the owner for each record."""
+
+    _object_type: str
+    """The name of the record class being encoded by this instance."""
+
+    _field_types: dict[str, type]
+    """A map of the fields on instances of the record class (and their types); these will appear in the object data."""
+
+    _has_failures: type[str | list[str]] | None
+    """The type of the failures attribute for this record, if present."""
+
+    _id_attribute_names: Sequence[str]
+    """The names of the record attributes that are used to produce the identifier for each record.
+
+    Attributes can be either a dataclass field or a property, for which the type must be a (non-optional) string.
+    """
+
+    def __init__(self, job_run_id: int, workspace_id: int, ownership: Ownership[Record], klass: type[Record]) -> None:
+        self._job_run_id = job_run_id
+        self._workspace_id = workspace_id
+        self._ownership = ownership
+        self._object_type = self._get_object_type(klass)
+        self._field_types, self._has_failures = self._get_field_types(klass)
```

Comment:
`self._field_types` suggests a list of types to me. Maybe introduce as small data class or named tuple `Field` with `(name, type)`

Reply:
I've renamed it… does that help?